src/rust/cryptography-x509-validation/src/policy/mod.rs

```diff
@@ -234,15 +249,340 @@ impl<'a, B: CryptoOps> Policy<'a, B> {
+    pub(crate) fn permits_leaf(
+        &self,
+        leaf: &Certificate<'_>,
+        extensions: &Extensions<'_>,
+    ) -> Result<(), ValidationError> {
+        // NOTE: Avoid refactoring this to `permits_ee() || permits_ca()` or any variation thereof.
+        // Code like this will propagate irrelevant error messages out of the API.
+        if let Some(key_usage) = extensions.get_extension(&KEY_USAGE_OID) {
+            let key_usage: KeyUsage<'_> = key_usage.value()?;
+            if key_usage.key_cert_sign() {
+                return self.permits_ca(leaf, 0, extensions);
+            }
+        }
+        self.permits_ee(leaf, extensions)
+    }
+
+    /// Checks whether the given CA certificate is compatible with this policy.
+    pub(crate) fn permits_ca(
+        &self,
+        cert: &Certificate<'_>,
+        current_depth: u8,
+        extensions: &Extensions<'_>,
+    ) -> Result<(), ValidationError> {
+        self.permits_basic(cert)?;
+
+        // 5280 4.1.2.6: Subject
+        // CA certificates MUST have a subject populated with a non-empty distinguished name.
+        // No check required here: `permits_basic` checks that the issuer is non-empty
+        // and `ChainBuilder::potential_issuers` enforces subject/issuer matching,
+        // meaning that an CA with an empty subject cannot occur in a built chain.
+
+        // NOTE: This conceptually belongs in `valid_issuer`, but is easier
+        // to test here. It's also conceptually an extension policy, but
+        // requires a bit of extra external state (`current_depth`) that isn't
+        // presently convenient to push into that layer.
+        if let Some(bc) = extensions.get_extension(&BASIC_CONSTRAINTS_OID) {
+            let bc: BasicConstraints = bc.value()?;
+
+            if bc
+                .path_length
+                .map_or(false, |len| current_depth as u64 > len)
+            {
+                return Err(ValidationError::Other(
+                    "path length constraint violated".to_string(),
+                ))?;
+            }
+        }
+
+        for ext_policy in self.ca_extension_policies.iter() {
+            ext_policy.permits(self, cert, extensions)?;
+        }
+
+        Ok(())
+    }
+
+    /// Checks whether the given EE certificate is compatible with this policy.
+    pub(crate) fn permits_ee(
+        &self,
+        cert: &Certificate<'_>,
+        extensions: &Extensions<'_>,
+    ) -> Result<(), ValidationError> {
+        self.permits_basic(cert)?;
+
+        for ext_policy in self.ee_extension_policies.iter() {
+            ext_policy.permits(self, cert, extensions)?;
+        }
+
+        Ok(())
+    }
+
+    /// Checks whether `issuer` is a valid issuing CA for `child` at a
+    /// path-building depth of `current_depth`.
+    ///
+    /// This checks that `issuer` is permitted under this policy and that
+    /// it was used to sign for `child`.
+    ///
```

Comment:
This should say that the caller should have already checked that issuer() == subject()

Reply:
SGTM. This is also making me realize that it's a little weird that we broke the `potential_issuers` and `valid_issuer` logic up like this -- I'll look into simplifying it.

Reply:
b1e8d2a5b